pkg/apis/cluster/common/plugins.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"github.com/golang/glog"
+	"sync"
+	"fmt"
+)
+
+var (
+	providersMutex sync.Mutex
+	providers      = make(map[string]interface{})
```

Comment:
why interface instead of using the `clusterdeployer.ProviderDeployer` type?

Reply:
`ProviderDeployer` is in clusterctl/clusterdeployer/clusterdeployer.go and i did not want pkg/apis/cluster/common/plugins.go to depend on the code from there. Also we need to get rid of ProviderDeployer so it's better to be left alone in clusterdeployer.go i think